src/pytorch_lightning/core/datamodule.py

```diff
@@ -150,7 +154,12 @@ def predict_dataloader():
                 return [dataloader(ds) for ds in predict_dataset]
             return dataloader(predict_dataset)
 
-        datamodule = cls()
+        candidate_kwargs = dict(batch_size=batch_size, num_workers=num_workers)
+        accepted_params = set(inspect.signature(cls.__init__).parameters.keys())
+        accepted_params.discard("self")
+        special_kwargs = {k: v for k, v in candidate_kwargs.items() if k in accepted_params}
```

Comment:
I would like to point out that this will not be passed if the DataModule has `*args, **kwargs`. Once again, we are battling the dragons that already exist in https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/utilities/data.py#L242 and that @otaj has been battling for a while with increasing complexity.

For this reason, I don't think I would do (2) as proposed in https://github.com/Lightning-AI/lightning/pull/14082#discussion_r944226325 and would instead go with option (3)

Reply:
If it's not too much to ask, could you @carmocca or @otaj directly implement your suggestion?
With the tests I added, I demonstrate that I address the problem in the linked issue. One can always iterate and generalize more.

@carmocca 
For the same reasons, I could also have said that we shouldn't move forward with #14212 because it's possible to come up with [edge cases where it breaks down](https://github.com/Lightning-AI/lightning/pull/14212#issuecomment-1215285742).

Reply:
Oh yeah, absolutely, I will implement it. Would you like me to hijack your PR or open a new one?

Reply:
it's already the 2nd PR for the same issue... let's hijack it 😄

Reply:
So, hijack is complete, I allowed to pass all the arguments when there is `kwargs` present. However, my second point stands moot, as I realized that since we are already passing `datamodule_kwargs` as kwargs, there is no way to have `batch_size` and `num_workers` in `datamodule_kwargs` as that call would already raise `TypeError` with multiple values

Reply:
Great! Thanks a lot